js/spreadsheet.js

```diff
@@ -67,248 +67,109 @@ var spreadsheet = function(options) {
-          break;
-        }
-      }
+    cellsGroup.forEach(function(cell) {
+      var hasData = tableData[cell[0]][cell[1]] !== null;
+      var processedRow = processedCells[cell[0]];
 
-      for (var i = startAt[0]; i > 0; i--) {
-        if (allData[i][lastCol] === null) {
-          firstRow = i ? i - 1 : i;
-        }
+      if (!processedRow) {
+        processedCells[cell[0]] = [];
       }
 
-      firstRow = firstRow || 0;
+      var processedCell = processedCells[cell[0]][cell[1]];
 
-      for (var i = firstRow; i < allData.length; i++) {
-        if (allData[i][lastCol] === null) {
-          lastRow = i - 1;
-          break;
-        }
+      if (!processedCell) {
+        processedCells[cell[0]][cell[1]] = false;
       }
-    } else if (moveTo.top) {
-      // When data located on the top of the selected cell
-      lastRow = startAt[0];
 
-      for (var i = lastRow - 1; i > 0; i--) {
-        if (allData[i][startAt[1]] === null) {
-          firstRow = i;
-          break;
-        }
-      }
-
-      firstRow = firstRow || 0;
-
-      for (var i = startAt[1]; i > 0; i--) {
-        if (allData[firstRow][i] === null) {
-          firstCol = i ? i + 1 : i;
-          break;
-        }
-      }
-
-      firstCol = firstCol || 0;
-
-      for (var i = firstCol; i < allData.length; i++) {
-        if (allData[firstRow][i] === null) {
-          lastCol = i - 1;
-          break;
-        }
-      }
-    } else if (moveTo.bottom) {
-      // When data located on the bottom of the selected cell
-      firstRow = startAt[0];
-
-      for (var i = firstRow + 1; i < allData.length; i++) {
-        if (allData[i][startAt[1]] === null) {
-          lastRow = i - 1;
-          break;
-        }
-      }
-
-      for (var i = startAt[1]; i > 0; i--) {
-        if (allData[lastRow][i] === null) {
-          firstCol = i + 1;
-          break;
-        }
+      if (hasData && (!processedRow || !processedCell)) {
+        processedCells[cell[0]][cell[1]] = true;
+        getDataCoordinates(tableData, cell, processedCells);
       }
+    });
 
-      firstCol = firstCol || 0;
+    return processedCells;
+  }
 
-      for (var i = firstCol; i < allData.length; i++) {
-        if (allData[lastRow][i] === null) {
-          lastCol = i - 1;
-          break;
-        }
-      }
-    } else if (moveTo.hasData) {
-      // When selected cell has data in it
-      if (startAt[1] === 0) {
-        firstCol = 0;
-      } else {
-        for (var i = startAt[1]; i > 0; i--) {
-          if (allData[startAt[0]][i] === null) {
-            firstCol = i + 1;
-            break;
-          }
+  /**
+   * The method that returns us a range that we should select in the table
+   * 
+   * @param {array} processedCells array processed by getDataCoordinates function
+   */
+  function getSelectionCoordinates(processedCells) {
+    var rowIndexes = [];
+    var colIndexes = [];
+
+    processedCells.forEach(function(row, rowIndex) {
```

Comment:
```suggestion

    processedCells.forEach(function(row, rowIndex) {
```

Reply:
@tonytlwu do we need one more new line here?

Reply:
No we don't :) looks like there's one already.